File: app/src/main/java/com/kota/Bahamut/Pages/BlockListPage/ArticleHeaderListPage.java
package com.kota.Bahamut.Pages.BlockListPage;

import static com.kota.Bahamut.Service.CommonFunctions.getContextString;
import static com.kota.Bahamut.Service.UserSettings.setBlockList;

import android.annotation.SuppressLint;
import android.util.Log;
import android.view.View;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.kota.ASFramework.Dialog.ASAlertDialog;
import com.kota.ASFramework.UI.ASToast;
import com.kota.Bahamut.BahamutPage;
import com.kota.Bahamut.Pages.PostArticlePage;
import com.kota.Bahamut.R;
import com.kota.Bahamut.Service.NotificationSettings;
import com.kota.Bahamut.Service.UserSettings;
import com.kota.TelnetUI.TelnetPage;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ArticleHeaderListPage extends TelnetPage implements BlockListClickListener {
    EditText _inputField;

    BlockListAdapter articleHeaderListAdapter;
    // 按下新增
    View.OnClickListener _addListener = v -> {
        if (_inputField != null) {
            String block_name = _inputField.getText().toString().trim();
            _inputField.setText("");
            if (block_name.length() > 0) {
                List<String> new_list = new ArrayList<>(Arrays.asList(UserSettings.getArticleHeaders()));
                if (new_list.contains(block_name)) {
                    ASAlertDialog.showErrorDialog(getContextString(R.string.already_have_item), ArticleHeaderListPage.this);
                } else {
                    new_list.add(block_name);
                }
                UserSettings.setArticleHeaders(new_list);

                UserSettings.notifyDataUpdated();
                ArticleHeaderListPage.this.reload();
            } else {
                ASAlertDialog.showErrorDialog(getContextString(R.string.please_input_id), ArticleHeaderListPage.this);
            }
        }
    };
    // 按下重置
    View.OnClickListener _resetListener = v -> {
        ASAlertDialog.createDialog()
                .setTitle(getContextString(R.string.reset))
                .setMessage(getContextString(R.string.reset_message))
                .addButton(getContextString(R.string.cancel))
                .addButton(getContextString(R.string.sure))
                .setListener((aDialog1, button_index) -> {
                    if (button_index > 0) {
                        ASToast.showShortToast(getContextString(R.string.reset_ok));
                        _inputField.setText("");
                        UserSettings.resetArticleHeaders();
                        ArticleHeaderListPage.this.reload();
                    }
                }).show();
    };
    ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP|ItemTouchHelper.DOWN,0) {
        public boolean isSwiped = false;
        public boolean isDragged = false;
        int start = -1;
        int end = -1;
        private View dragView;
        // 上下移動
        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
            start = viewHolder.getAdapterPosition();
            end = target.getAdapterPosition();
                if (UserSettings.getPropertiesVIP()) {
                    if (start==0 || end==0) {
                        ASToast.showShortToast(getContextString(R.string.article_header_zero_error_message));
                    } else {
                        Collections.swap(_articleHeadersList, start, end);
                        articleHeaderListAdapter.notifyItemMoved(start, end);
                    }
                } else {
                    ASToast.showShortToast(getContextString(R.string.vip_only_message));
                }
            return true;
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {

        }

        @Override
        public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
            switch (actionState) {
                case ItemTouchHelper.ACTION_STATE_DRAG -> {
                    // the user is dragging an item and didn't lift their finger off yet
                    isSwiped = false;
                    isDragged = true;
                    if (viewHolder != null) { // 選取變色
                        dragView = viewHolder.itemView;
                        dragView.setBackgroundResource(R.color.ripple_material);
                    }
                }
                case ItemTouchHelper.ACTION_STATE_SWIPE -> {
                    // the user is swiping an item and didn't lift their finger off yet
                    isSwiped = true;
                    isDragged = false;
                }
                case ItemTouchHelper.ACTION_STATE_IDLE -> {
                    // the user just dropped the item (after dragging it), and lift their finger off.
                    //
                    if (isSwiped) { // The user used onSwiped()
                        Log.e(getClass().getSimpleName(), "swipe is over");
                    }
                    if (!isSwiped && isDragged) { // The user used onMove()
                        if (dragView != null) { // 解除 選取變色
                            dragView.setBackgroundResource(R.color.transparent);
                            dragView = null;
                        }
                        UserSettings.setArticleHeaders(_articleHeadersList);
                    }
                    isSwiped = false;
                    isDragged = false;
                }
            }
        }
    });


    private List<String> _articleHeadersList = new ArrayList<>();

    public int getPageType() {
        return BahamutPage.BAHAMUT_BLOCK_LIST;
    }

    public int getPageLayout() {
        return R.layout.block_list_page;
    }

    public boolean isKeepOnOffline() {
        return true;
    }

    public void onPageDidLoad() {
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.BlockList_list);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        itemTouchHelper.attachToRecyclerView(recyclerView);

        articleHeaderListAdapter = new BlockListAdapter(_articleHeadersList);
        recyclerView.setAdapter(articleHeaderListAdapter);
        articleHeaderListAdapter.setOnItemClickListener(this);

        _inputField = (EditText) findViewById(R.id.BlockList_Input);
        findViewById(R.id.BlockList_Add).setOnClickListener(_addListener);
        findViewById(R.id.BlockList_Reset).setOnClickListener(_resetListener);

        showNotification();

        reload();
    }

    // 第一次進入的提示訊息
    private void showNotification() {
        boolean show_top_bottom_function = NotificationSettings.getShowHeader();
        if (!show_top_bottom_function) {
            ASToast.showLongToast(getContextString(R.string.notification_header));
            NotificationSettings.setShowHeader(true);
        }
    }

    public void onPageWillDisappear() {
        UserSettings.notifyDataUpdated();
    }

    public void onPageDidDisappear() {
        _articleHeadersList = null;
        super.onPageDidDisappear();
    }

    @SuppressLint("NotifyDataSetChanged")
    private void reload() {
        List<String> _temp = Arrays.asList(UserSettings.getArticleHeaders());
        _articleHeadersList.clear();
        _articleHeadersList.addAll(_temp);
        articleHeaderListAdapter.notifyDataSetChanged();
    }

    public boolean onReceivedGestureRight() {
        onBackPressed();
        ASToast.showShortToast("返回");
        return true;
    }

    @Override
    public void onBlockListPage_ItemView_clicked(BlockListViewHolder blockListPage_ItemView) {

    }

    // 刪除黑名單
    @Override
    public void onBlockListPage_ItemView_delete_clicked(BlockListViewHolder blockListPage_ItemView) {
        int deleted_index = blockListPage_ItemView.getAdapterPosition();
        if (deleted_index==0) {
            ASToast.showShortToast(getContextString(R.string.article_header_zero_error_message));
            return;
        }
        List<String> new_list = ArticleHeaderListPage.this._articleHeadersList;
        new_list.remove(deleted_index);

        // 更新
        UserSettings.setArticleHeaders(_articleHeadersList);
        ArticleHeaderListPage.this.reload();
    }
}

File: app/src/main/java/com/kota/Bahamut/pages/blockListPage/ArticleHeaderListPage.kt
package com.kota.Bahamut.pages.blockListPage

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import android.view.View
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.HorizontalDivider
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.kota.Bahamut.ui.components.BahaInputField
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.kota.Bahamut.BahamutPage
import com.kota.Bahamut.R
import com.kota.Bahamut.service.CommonFunctions.getContextString
import com.kota.Bahamut.service.NotificationSettings.getShowHeader
import com.kota.Bahamut.service.NotificationSettings.setShowHeader
import com.kota.Bahamut.service.UserSettings
import com.kota.Bahamut.service.UserSettings.Companion.articleHeaders
import com.kota.Bahamut.service.UserSettings.Companion.notifyDataUpdated
import com.kota.Bahamut.service.UserSettings.Companion.propertiesVIP
import com.kota.Bahamut.service.UserSettings.Companion.resetArticleHeaders
import com.kota.Bahamut.ui.components.BahaButton
import com.kota.Bahamut.ui.components.ButtonType
import com.kota.Bahamut.ui.dialogs.BahaGlobalDialogHost
import com.kota.Bahamut.ui.theme.AppTheme
import com.kota.Bahamut.ui.theme.setBahamutContent
import com.kota.asFramework.dialog.ASAlertDialog
import com.kota.asFramework.dialog.ASAlertDialog.Companion.createDialog
import com.kota.asFramework.dialog.ASAlertDialog.Companion.showErrorDialog
import com.kota.asFramework.ui.ASToast.showLongToast
import com.kota.asFramework.ui.ASToast.showShortToast
import com.kota.telnetUI.TelnetPage
import java.util.Collections

class ArticleHeaderListPage : TelnetPage(), BlockListClickListener {

    var inputTextState by mutableStateOf("")

    private var articleHeaderListAdapter: BlockListAdapter? = null
    private var articleHeadersList: MutableList<String> = ArrayList()

    override val isPopupPage: Boolean
        get() = true

    override val pageType: Int
        get() = BahamutPage.BAHAMUT_BLOCK_LIST

    override val pageLayout: Int
        get() = 0

    override val isKeepOnOffline: Boolean
        get() = true

    val itemTouchHelper: ItemTouchHelper = ItemTouchHelper(object :
        ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, 0) {
        var isSwiped: Boolean = false
        var isDragged: Boolean = false
        private var dragView: View? = null

        override fun onMove(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder,
            target: RecyclerView.ViewHolder
        ): Boolean {
            val start = viewHolder.bindingAdapterPosition
            val end = target.bindingAdapterPosition
            if (propertiesVIP) {
                if (start == 0 || end == 0) {
                    showShortToast(getContextString(R.string.article_header_zero_error_message))
                } else {
                    Collections.swap(articleHeadersList, start, end)
                    articleHeaderListAdapter?.notifyItemMoved(start, end)
                }
            } else {
                showShortToast(getContextString(R.string.vip_only_message))
            }
            return true
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}

        override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
            when (actionState) {
                ItemTouchHelper.ACTION_STATE_DRAG -> {
                    isSwiped = false; isDragged = true
                    dragView = viewHolder?.itemView
                    dragView?.setBackgroundResource(R.color.ripple_material)
                }
                ItemTouchHelper.ACTION_STATE_SWIPE -> { isSwiped = true; isDragged = false }
                ItemTouchHelper.ACTION_STATE_IDLE -> {
                    if (isSwiped) Log.e("swipe", "swipe is over")
                    if (!isSwiped && isDragged) {
                        dragView?.setBackgroundResource(R.color.transparent)
                        dragView = null
                        UserSettings.setArticleHeaders(articleHeadersList)
                    }
                    isSwiped = false; isDragged = false
                }
            }
        }
    })

    override fun createPageView(context: Context): View {
        return ComposeView(context).apply {
            setBahamutContent {
                ArticleHeaderListPageContent()
                BahaGlobalDialogHost()
            }
        }
    }

    override fun onPageWillDisappear() {
        notifyDataUpdated()
    }

    override fun onPageDidDisappear() {
        articleHeadersList = ArrayList()
        super.onPageDidDisappear()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun reload() {
        val elements = listOf(*articleHeaders)
        articleHeadersList.clear()
        articleHeadersList.addAll(elements)
        articleHeaderListAdapter?.notifyDataSetChanged()
    }

    private fun showNotification() {
        if (!getShowHeader()) {
            showLongToast(getContextString(R.string.notification_header))
            setShowHeader(true)
        }
    }

    private fun onAddClicked() {
        val name = inputTextState.trim()
        inputTextState = ""
        if (name.isNotEmpty()) {
            val newList: MutableList<String> = ArrayList(listOf(*articleHeaders))
            if (newList.contains(name)) {
                showErrorDialog(getContextString(R.string.already_have_item), this)
            } else {
                newList.add(name)
            }
            UserSettings.setArticleHeaders(newList)
            notifyDataUpdated()
            reload()
        } else {
            showErrorDialog(getContextString(R.string.please_input_id), this)
        }
    }

    private fun onResetClicked() {
        createDialog()
            .setTitle(getContextString(R.string.reset))
            .setMessage(getContextString(R.string.reset_message))
            .addButton(getContextString(R.string.cancel))
            .addButton(getContextString(R.string.sure))
            .setListener { _: ASAlertDialog?, buttonIndex: Int ->
                if (buttonIndex > 0) {
                    showShortToast(getContextString(R.string.reset_ok))
                    inputTextState = ""
                    resetArticleHeaders()
                    reload()
                }
            }.show()
    }

    override fun onReceivedGestureRight(): Boolean {
        onBackPressed()
        showShortToast("返回")
        return true
    }

    override fun onBlockListPageItemViewClicked(blockListPageItemView: BlockListViewHolder) {}

    override fun onBlockListPageItemViewDeleteClicked(blockListPageItemView: BlockListViewHolder) {
        val deletedIndex = blockListPageItemView.bindingAdapterPosition
        if (deletedIndex == 0) {
            showShortToast(getContextString(R.string.article_header_zero_error_message))
            return
        }
        articleHeadersList.removeAt(deletedIndex)
        UserSettings.setArticleHeaders(articleHeadersList)
        reload()
    }

    @Composable
    fun ArticleHeaderListPageContent() {
        val colors = AppTheme.colors

        if (articleHeaderListAdapter == null) {
            articleHeaderListAdapter = BlockListAdapter(articleHeadersList).also {
                it.setOnItemClickListener(this)
            }
            showNotification()
            reload()
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .background(colors.pageBackground)
        ) {
            HorizontalDivider(color = colors.divider, thickness = 1.dp)
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(60.dp)
                    .background(colors.pageBackground),
                verticalAlignment = Alignment.CenterVertically
            ) {
                BahaButton(
                    text = stringResource(R.string.reset),
                    type = ButtonType.DANGER,
                    modifier = Modifier.fillMaxHeight(),
                    onClick = { onResetClicked() }
                )
                BahaInputField(
                    value = inputTextState,
                    onValueChange = { inputTextState = it },
                    placeholder = stringResource(R.string.please_input_id),
                    modifier = Modifier
                        .weight(1f)
                        .padding(horizontal = 8.dp),
                    keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                    keyboardActions = KeyboardActions(onDone = { onAddClicked() })
                )
                BahaButton(
                    text = stringResource(R.string.add),
                    modifier = Modifier.fillMaxHeight(),
                    onClick = { onAddClicked() }
                )
            }
            HorizontalDivider(color = colors.divider, thickness = 1.dp)

            AndroidView(
                factory = { ctx ->
                    RecyclerView(ctx).apply {
                        layoutManager = LinearLayoutManager(ctx)
                        adapter = articleHeaderListAdapter
                        itemTouchHelper.attachToRecyclerView(this)
                    }
                },
                modifier = Modifier.weight(1f).fillMaxWidth()
            )

            // 底部工具列 (滿版無縫)
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(colors.toolbarDivider)
            )
            BahaButton(
                text = stringResource(R.string._back),
                modifier = Modifier
                    .fillMaxWidth()
                    .height(60.dp),
                onClick = { onBackPressed() }
            )
        }
    }
}
